### PR TITLE
feat: Logical filter cache

### PR DESCRIPTION
**Pull request by dungnmaster**
- added changes to enable caching of UDF results when they are present in Logical filters

### CODE REVIEW

File: eva/expression/function_expression.py [hunk 1]
```diff
@@ -210,7 +210,7 @@ def _apply_function_expression(self, func: Callable, batch: Batch, **kwargs):
         cache_miss = np.full(len(batch), True)
         for idx, key in cache_keys.iterrows():
             val = self._cache.store.get(key.to_numpy())
-            results[idx] = val
+            results[idx] = val[output_cols].to_numpy() if val is not None else val
```
Comment (gaurav274) [inline eva/expression/function_expression.py:213]: Why do we need `output_cols` here?

File: eva/expression/function_expression.py [hunk 2]
```diff
@@ -226,10 +226,12 @@ def _apply_function_expression(self, func: Callable, batch: Batch, **kwargs):
             for key, value in zip(
                 missing_keys.iterrows(), cache_miss_results.iterrows()
             ):
-                self._cache.store.set(key[1].to_numpy(), value[1].to_numpy())
+                self._cache.store.set(key[1].to_numpy(), value[1])
```
Comment (gaurav274) [inline eva/expression/function_expression.py:229]: Same here.

File: eva/optimizer/optimizer_utils.py [hunk 3]
```diff
@@ -239,19 +255,25 @@ def enable_cache(func_expr: FunctionExpression) -> FunctionExpression:
...
-    if not cache_entry:
-        cache_entry = CatalogManager().insert_udf_cache_catalog_entry(func_expr)
 
-    cache = FunctionExpressionCache(
-        key=tuple(optimized_key), store=DiskKVCache(cache_entry.cache_path)
+def enable_cache_on_tree(expr_tree: AbstractExpression):
+    func_exprs = list(expr_tree.find_all(FunctionExpression))
+    func_exprs = list(filter(lambda expr: check_expr_validity(expr), func_exprs))
+    for expr in func_exprs:
+        cache = enable_cache_init(expr)
```
Comment (gaurav274) [inline eva/optimizer/optimizer_utils.py:266]: Please change the function name. How about `enable_cache_on_function_expression`?

File: eva/optimizer/optimizer_utils.py [hunk 4]
```diff
@@ -239,19 +255,25 @@ def enable_cache(func_expr: FunctionExpression) -> FunctionExpression:
...
+        expr.enable_cache(cache)
+
+
+def check_expr_validity(expr: FunctionExpression):
```
Comment (gaurav274) [inline eva/optimizer/optimizer_utils.py:270]: Please add doc strings to the functions. Thanks!

File: eva/optimizer/rules/rules.py [hunk 5]
```diff
@@ -186,6 +188,37 @@ def apply(self, before: LogicalSample, context: OptimizerContext):
...
+        pattern.append_child(Pattern(OperatorType.DUMMY))
+        super().__init__(RuleType.CACHE_FUNCTION_EXPRESISON_IN_FILTER, pattern)
+
+    def promise(self):
+        return Promise.CACHE_FUNCTION_EXPRESISON_IN_FILTER
+
+    def check(self, before: LogicalFilter, context: OptimizerContext):
+        func_exprs = list(before.predicate.find_all(FunctionExpression))
+
+        for temp in func_exprs:
```
Comment (gaurav274) [inline eva/optimizer/rules/rules.py:203]: We can remove this.

File: test/integration_tests/test_reuse.py [hunk 6]
```diff
@@ -109,6 +119,10 @@ def test_reuse_does_not_work_when_expression_in_where_clause(self):
         select_query = """
             SELECT id FROM DETRAC
             WHERE ArrayCount(YoloV5(data).labels, 'car') > 3 AND id < 5;"""
+
```
Comment (gaurav274) [inline test/integration_tests/test_reuse.py:122]: This should work now, right?

Comment (gaurav274) [reply test/integration_tests/test_reuse.py:122]: Similarly, look at the query on line 112. That should also work now. Please enable it also. Thanks!

**Review by gaurav274 — COMMENTED** [review]
The implementation is spot on! Dropped minor comments. Thanks for working on it.

File: eva/expression/function_expression.py (hunk 1 above)
Comment (dungnmaster) [reply eva/expression/function_expression.py:213]: there is a shape mismatch when the output_cols is a subset of the fields returned by the model

File: eva/optimizer/optimizer_utils.py (hunk 3 above)
Comment (dungnmaster) [reply eva/optimizer/optimizer_utils.py:266]: will change to `enable_cache_on_function_expression`

**Pull request merged**